File: services/dotnet-microservices/UserService/Application/UserServiceImpl.cs
namespace UserService.Application;

using System.Security.Cryptography;
using System.Text;
using UserService.Domain.Interfaces;
using UserService.Domain.Models;
using UserService.Infrastructure.Messaging;

public class UserServiceImpl : IUserService
{
    private readonly IUserRepository _repository;
    private readonly KafkaProducer _kafkaProducer;
    private readonly ILogger<UserServiceImpl> _logger;

    public UserServiceImpl(IUserRepository repository, KafkaProducer kafkaProducer, ILogger<UserServiceImpl> logger)
    {
        _repository = repository;
        _kafkaProducer = kafkaProducer;
        _logger = logger;
    }

    public async Task<User?> GetUserByIdAsync(Guid id) => await _repository.GetByIdAsync(id);

    public async Task<User?> GetUserByEmailAsync(string email) => await _repository.GetByEmailAsync(email);

    public async Task<IEnumerable<User>> GetAllUsersAsync() => await _repository.GetAllAsync();

    public async Task<User> CreateUserAsync(CreateUserRequest request)
    {
        var existing = await _repository.GetByEmailAsync(request.Email);
        if (existing != null)
            throw new InvalidOperationException($"User with email {request.Email} already exists");

        var user = new User(
            Id: Guid.NewGuid(),
            Email: request.Email,
            FirstName: request.FirstName,
            LastName: request.LastName,
            PasswordHash: HashPassword(request.Password),
            Phone: request.Phone,
            IsActive: true,
            CreatedAt: DateTime.UtcNow,
            UpdatedAt: DateTime.UtcNow,
            Addresses: new List<Address>()
        );

        var created = await _repository.CreateAsync(user);

        await _kafkaProducer.PublishAsync("user-events", new
        {
            EventType = "UserCreated",
            UserId = created.Id,
            Email = created.Email,
            Timestamp = DateTime.UtcNow
        });

        _logger.LogInformation("User created: {UserId}", created.Id);
        return created;
    }

    public async Task<User> UpdateUserAsync(Guid id, UpdateUserRequest request)
    {
        var user = await _repository.GetByIdAsync(id)
            ?? throw new KeyNotFoundException($"User {id} not found");

        var updated = user with
        {
            FirstName = request.FirstName ?? user.FirstName,
            LastName = request.LastName ?? user.LastName,
            Phone = request.Phone ?? user.Phone,
            UpdatedAt = DateTime.UtcNow
        };

        var result = await _repository.UpdateAsync(updated);

        await _kafkaProducer.PublishAsync("user-events", new
        {
            EventType = "UserUpdated",
            UserId = result.Id,
            Timestamp = DateTime.UtcNow
        });

        return result;
    }

    public async Task<bool> DeleteUserAsync(Guid id)
    {
        var deleted = await _repository.DeleteAsync(id);
        if (deleted)
        {
            await _kafkaProducer.PublishAsync("user-events", new
            {
                EventType = "UserDeleted",
                UserId = id,
                Timestamp = DateTime.UtcNow
            });
        }
        return deleted;
    }

    public async Task<Address> AddAddressAsync(Guid userId, CreateAddressRequest request)
    {
        var user = await _repository.GetByIdAsync(userId)
            ?? throw new KeyNotFoundException($"User {userId} not found");

        var address = new Address(
            Id: Guid.NewGuid(),
            UserId: userId,
            Street: request.Street,
            City: request.City,
            State: request.State,
            ZipCode: request.ZipCode,
            Country: request.Country,
            IsDefault: request.IsDefault,
            CreatedAt: DateTime.UtcNow
        );

        return await _repository.AddAddressAsync(address);
    }

    public async Task<IEnumerable<Address>> GetUserAddressesAsync(Guid userId)
    {
        return await _repository.GetAddressesByUserIdAsync(userId);
    }

    private static string HashPassword(string password)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(password));
        return Convert.ToBase64String(bytes);
    }
}

File: services/nodejs-microservices/user-service/src/application/UserServiceImpl.ts
/**
 * User Service Implementation
 * 
 * Contains the business logic for user operations.
 * This is the Application Layer - it orchestrates domain objects
 * and infrastructure services to fulfill use cases.
 * 
 * Pattern: Service Layer
 * - Encapsulates business logic
 * - Coordinates between repository and messaging
 * - Publishes domain events after state changes
 */
import { v4 as uuidv4 } from 'uuid';
import { User, Address, UserStatus } from '../domain/models/User';
import { IUserService, CreateUserDTO, UpdateUserDTO, CreateAddressDTO } from '../domain/interfaces/IUserService';
import { UserRepository } from '../infrastructure/persistence/UserRepository';
import { KafkaProducer } from '../infrastructure/messaging/KafkaProducer';
import { USER_EVENTS_TOPIC, UserEventType } from '../infrastructure/messaging/events';

export class UserServiceImpl implements IUserService {
  constructor(
    private readonly userRepository: UserRepository,
    private readonly kafkaProducer: KafkaProducer
  ) {}

  async createUser(dto: CreateUserDTO): Promise<User> {
    // Check for duplicate email
    const existing = await this.userRepository.findByEmail(dto.email);
    if (existing) {
      throw new Error('User with this email already exists');
    }

    // In production, use bcrypt for password hashing
    const passwordHash = Buffer.from(dto.password).toString('base64');

    const user: Omit<User, 'createdAt' | 'updatedAt'> = {
      id: uuidv4(),
      email: dto.email,
      passwordHash,
      firstName: dto.firstName,
      lastName: dto.lastName,
      phone: dto.phone,
      status: UserStatus.ACTIVE,
    };

    const created = await this.userRepository.create(user);

    // Publish domain event - other services react to this
    await this.kafkaProducer.publish(USER_EVENTS_TOPIC, {
      type: UserEventType.USER_CREATED,
      data: { id: created.id, email: created.email, firstName: created.firstName, lastName: created.lastName },
    });

    return created;
  }

  async getUserById(id: string): Promise<User | null> {
    return this.userRepository.findById(id);
  }

  async getUserByEmail(email: string): Promise<User | null> {
    return this.userRepository.findByEmail(email);
  }

  async updateUser(id: string, dto: UpdateUserDTO): Promise<User | null> {
    const existing = await this.userRepository.findById(id);
    if (!existing) return null;

    const updated = await this.userRepository.update(id, {
      firstName: dto.firstName,
      lastName: dto.lastName,
      phone: dto.phone,
      status: dto.status as UserStatus,
    });

    if (updated) {
      await this.kafkaProducer.publish(USER_EVENTS_TOPIC, {
        type: UserEventType.USER_UPDATED,
        data: { id: updated.id, email: updated.email, firstName: updated.firstName, lastName: updated.lastName },
      });
    }

    return updated;
  }

  async deleteUser(id: string): Promise<boolean> {
    const deleted = await this.userRepository.delete(id);
    if (deleted) {
      await this.kafkaProducer.publish(USER_EVENTS_TOPIC, {
        type: UserEventType.USER_DELETED,
        data: { id },
      });
    }
    return deleted;
  }

  async listUsers(page: number, limit: number): Promise<User[]> {
    const offset = (page - 1) * limit;
    return this.userRepository.findAll(offset, limit);
  }

  async addAddress(userId: string, dto: CreateAddressDTO): Promise<Address> {
    // Verify user exists
    const user = await this.userRepository.findById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    const address: Omit<Address, 'createdAt'> = {
      id: uuidv4(),
      userId,
      street: dto.street,
      city: dto.city,
      state: dto.state,
      zipCode: dto.zipCode,
      country: dto.country,
      isDefault: dto.isDefault || false,
    };

    const created = await this.userRepository.createAddress(address);

    await this.kafkaProducer.publish(USER_EVENTS_TOPIC, {
      type: UserEventType.ADDRESS_ADDED,
      data: { userId, addressId: created.id },
    });

    return created;
  }

  async getAddresses(userId: string): Promise<Address[]> {
    return this.userRepository.findAddressesByUserId(userId);
  }

  async deleteAddress(userId: string, addressId: string): Promise<boolean> {
    return this.userRepository.deleteAddress(userId, addressId);
  }
}
